File: scripts/current_host_review_cleanup.mjs
/** Fail-closed repository-local editorial cleanup above the jurisdiction gate. */
import crypto from 'node:crypto';
import {loadJurisdiction} from './current_host_jurisdiction.mjs';

export const CLEANUP_PATH='verification/current-host-review-cleanup.json';
// Filled with the sealed registry digest when the independent proposal freezes.
export const CLEANUP_SHA256='ae2c8781c9d0d806818649052ad9bace4e8e6e9cf0b08e61f1d797049c20906c';
export const CLEANUP_ATTEMPT='verification/evidence/2026-09-11-host-review-cleanup-attempt-01';
export const CLEANUP_BASELINE=CLEANUP_ATTEMPT+'/before-current-host-docs-review.json';
export const CLEANUP_BASELINE_SHA256='f18e61882f50a2785f2fe863aba88ecf3d79af8a45bdcbf1596e6e64c6b7059d';
export const CLEANUP_MARKER='HOST-REVIEW-CLEANUP-01';
export const CLEANUP_DECISION='CURRENT_HOST_REVIEW_CLEANUP';
const STATIC='REPOSITORY_STATIC_CHECK';
const EDITORIAL=new Set(['EDITORIAL_SCOPE_OR_LEAD_IN','EDITORIAL_NAVIGATION_PREAMBLE']);
const STATIC_CLASSES=new Set(['REVIEWED_ADVICE','REVIEWED_NAVIGATION_INSTRUCTION','REVIEWED_EXAMPLE_CALCULATION']);
const METHODS=new Set(['MARKUP_OR_PREAMBLE_INSPECTION','CONTEXTUAL_INSPECTION','LOCAL_NAVIGATION_CHECK','ARITHMETIC_CHECK','MIXED_REPOSITORY_LOCAL_EDITORIAL_CHECKS']);
const AFTER=['classification','status','required_evidence_types','owner_role','rationale','next_action','evidence_refs','source_refs'];
const hash=x=>crypto.createHash('sha256').update(x).digest('hex');
const req=(x,message)=>{if(!x)throw Error('cleanup: '+message)};
const canon=x=>Array.isArray(x)?x.map(canon):x&&typeof x==='object'?Object.fromEntries(Object.keys(x).sort().map(k=>[k,canon(x[k])])):x;
const equal=(a,b)=>JSON.stringify(canon(a))===JSON.stringify(canon(b));
const objhash=x=>hash(JSON.stringify(canon(x)));
const keys=(x,names)=>req(x&&typeof x==='object'&&!Array.isArray(x)&&equal(Object.keys(x).sort(),[...names].sort()),'missing/unknown fields');
const index=model=>new Map(model.pages.flatMap(page=>page.claims.map(claim=>[claim.id,claim])));

function correction(count){return {id:CLEANUP_MARKER,scope:`${count} exact editorial/navigation/calculation cleanup transitions`,
 history:'The sealed jurisdiction predecessor is hash-pinned. Original claim text, spans, statuses, sources and evidence remain available.',
 current:`${count} individually observed repository-local editorial checks; no Host, API, SSH, credentialed, paid or mutating operation was run.`,
 reason:'A completed navigation, arithmetic or contextual editorial inspection is limited to the documented passage. It does not establish product behavior, a runtime result, policy enforcement or human acceptance.'};}

export function projectReviewCleanup({read,exists}){
 const pinned=(ref,wanted)=>{req(typeof ref==='string'&&!ref.startsWith('/')&&!ref.includes('\\')&&ref.split('/').every(p=>p&&!['.','..'].includes(p)),'unsafe path');const value=read(ref);req(hash(value)===wanted,'digest drift '+ref);return value};
 req(CLEANUP_SHA256,'cleanup registry digest not sealed');const registryBytes=pinned(CLEANUP_PATH,CLEANUP_SHA256),registrySha256=CLEANUP_SHA256,registry=JSON.parse(registryBytes);
 keys(registry,['schema_version','record_type','generated_at','baseline','artifacts','transitions','result_ref']);
 req(registry.schema_version==='1.0'&&registry.record_type==='HOST_REVIEW_CLEANUP_TRANSITION','wrong registry type');
 req(equal(registry.baseline,{path:CLEANUP_BASELINE,sha256:CLEANUP_BASELINE_SHA256}),'baseline substitution');
 req(typeof registry.result_ref==='string'&&registry.result_ref.startsWith(CLEANUP_ATTEMPT+'/')&&(exists?exists(registry.result_ref):!!read(registry.result_ref)),'invalid result reference');
 const baseline=JSON.parse(pinned(CLEANUP_BASELINE,CLEANUP_BASELINE_SHA256));
 const predecessor=loadJurisdiction({read,model:baseline,exists});req(predecessor?.jurisdiction,'sealed jurisdiction predecessor required');
 const artifacts=new Map();
 for(const artifact of registry.artifacts){
  keys(artifact,['id','path','sha256','method']);const id=artifact.id;
  req(typeof id==='string'&&id&&!artifacts.has(id)&&METHODS.has(artifact.method),'invalid artifact');req(typeof artifact.path==='string'&&artifact.path.startsWith(CLEANUP_ATTEMPT+'/'),'artifact outside cleanup attempt');
  const capture=JSON.parse(pinned(artifact.path,artifact.sha256));keys(capture,['schema_version','record_type','generated_at','baseline','method','limits','source_candidate','support_artifacts','observations']);
  req(capture.schema_version==='1.0'&&capture.record_type==='HOST_REVIEW_CLEANUP_LOCAL_CHECKS','wrong cleanup capture');req(equal(capture.baseline,registry.baseline)&&capture.method===artifact.method,'capture baseline/method drift');req(typeof capture.limits==='string'&&capture.limits.trim()&&Array.isArray(capture.observations)&&Array.isArray(capture.support_artifacts),'capture limits/observations required');keys(capture.source_candidate,['path','sha256']);const candidate=JSON.parse(pinned(capture.source_candidate.path,capture.source_candidate.sha256));req(candidate.record_type==='HOST_EDITORIAL_INSPECTION_CANDIDATES'&&equal(candidate.baseline,registry.baseline)&&Array.isArray(candidate.records),'raw independent candidate drift');for(const support of capture.support_artifacts){keys(support,['path','sha256']);pinned(support.path,support.sha256)}artifacts.set(id,{artifact,capture,candidates:new Map(candidate.records.map(item=>[item.claim_id,item]))});
 }
 const model=structuredClone(baseline),before=index(baseline),after=index(model),seen=new Set(),presentation=new Map(predecessor.presentation);
 for(const entry of registry.transitions){
  keys(entry,['claim_id','before_sha256','artifact_id','observation_id','after','review_rationale','limits']);const id=entry.claim_id;
  req(before.has(id)&&!seen.has(id)&&entry.before_sha256===objhash(before.get(id)),'claim identity/predecessor drift');seen.add(id);
  const prior=before.get(id),claim=after.get(id);req(prior.status==='UNVALIDATED','only known selected UNVALIDATED claims may transition');keys(entry.after,AFTER);Object.assign(claim,structuredClone(entry.after));
  req(['PASS','NOT_APPLICABLE','UNVALIDATED'].includes(claim.status),'only PASS, NOT_APPLICABLE or scoped UNVALIDATED cleanup status');req(claim.text===prior.text&&equal(claim.headings,prior.headings)&&equal(claim.spans,prior.spans),'source literal/span mutation');
  req(prior.evidence_refs.every(ref=>claim.evidence_refs.some(item=>equal(item,ref)))&&prior.source_refs.every(ref=>claim.source_refs.some(item=>equal(item,ref))),'prior evidence/source dropped');
  const bundle=artifacts.get(entry.artifact_id);req(bundle,'missing cleanup artifact');const observation=bundle.capture.observations.find(item=>item.id===entry.observation_id);
  keys(observation,['id','claim_id','before_sha256','text','text_sha256','source_file','source_sha256','spans','method','result','finding','limits','context','expected','observed','links','arithmetic','candidate_record']);req(prior.spans.length&&prior.spans.every(span=>span.source_file===prior.spans[0].source_file),'cleanup spans must be in one exact source file');const sourceFile=prior.spans[0].source_file;
  req(observation.claim_id===id&&observation.before_sha256===entry.before_sha256&&observation.text===prior.text&&observation.text_sha256===hash(prior.text),'observation claim identity drift');
  req(observation.source_file===sourceFile&&equal(observation.spans,prior.spans)&&observation.source_sha256===hash(read(sourceFile)),'observation source/span drift');req(METHODS.has(observation.method)&&bundle.artifact.method==='MIXED_REPOSITORY_LOCAL_EDITORIAL_CHECKS'&&observation.finding===entry.review_rationale&&observation.limits===entry.limits,'observation rationale/limits drift');req(equal(observation.candidate_record,bundle.candidates.get(id)),'raw candidate record substitution');req(typeof observation.context==='string'&&read(sourceFile).toString().includes(observation.context)&&typeof observation.expected==='string'&&observation.expected.trim()&&typeof observation.observed==='string'&&observation.observed.trim(),'raw contextual expected/observed evidence missing');req(Array.isArray(observation.links),'navigation evidence must be a list');req(typeof entry.review_rationale==='string'&&entry.review_rationale.trim()&&typeof entry.limits==='string'&&entry.limits.trim(),'rationale/limits required');
  if(claim.status==='NOT_APPLICABLE')req(EDITORIAL.has(claim.classification)&&equal(claim.required_evidence_types,[])&&observation.method==='MARKUP_OR_PREAMBLE_INSPECTION'&&observation.result==='NOT_APPLICABLE'&&observation.arithmetic===null,'invalid editorial non-claim closure');
  else {req(STATIC_CLASSES.has(claim.classification)&&equal(claim.required_evidence_types,[STATIC])&&['CONTEXTUAL_INSPECTION','LOCAL_NAVIGATION_CHECK','ARITHMETIC_CHECK'].includes(observation.method)&&observation.result===(claim.status==='PASS'?'PASS':'UNVALIDATED'),'invalid bounded static closure');req(claim.evidence_refs.some(ref=>ref.artifact_ref===bundle.artifact.path),'selected static evidence missing');}
  if(claim.status==='PASS'&&observation.method==='LOCAL_NAVIGATION_CHECK')req(observation.links.length&&observation.arithmetic===null&&observation.links.every(item=>item&&typeof item==='object'&&equal(Object.keys(item).sort(),['destination','href','result'])&&item.result==='PASS'),'exact local navigation observation required');
  if(claim.status==='PASS'&&observation.method==='ARITHMETIC_CHECK')req(!observation.links.length&&observation.arithmetic&&typeof observation.arithmetic==='object'&&equal(Object.keys(observation.arithmetic).sort(),['expected','formula','inputs','observed'])&&observation.arithmetic.expected===observation.arithmetic.observed,'exact arithmetic observation required');
  if(observation.method==='CONTEXTUAL_INSPECTION')req(!observation.links.length&&observation.arithmetic===null,'contextual inspection must not pose as link/arithmetic proof');
  if(claim.status==='UNVALIDATED')req(observation.method==='CONTEXTUAL_INSPECTION'&&claim.rationale!==prior.rationale&&claim.next_action!==prior.next_action,'unvalidated cleanup needs exact current rationale and next action');
  claim.history={...prior.history,carry_decision:CLEANUP_DECISION,reason:(prior.history.reason||'')+' Exact repository-local cleanup observation; no product/runtime or human-acceptance result is inferred.'};claim.coverage_state='CHANGED';
  const earlier=presentation.get(id),observationIndex=bundle.capture.observations.indexOf(observation);presentation.set(id,{previous:structuredClone(prior),method:observation.method,reviewRationale:entry.review_rationale,remaining:entry.limits+' '+claim.next_action,registryRef:CLEANUP_PATH,baselineRef:CLEANUP_BASELINE,basis:[{kind:'DOCUMENTATION_CHECK',sourceLabel:'Documentation check',sourceLocator:'Observation '+observation.id,sourceUrl:null,artifactRef:bundle.artifact.path,method:observation.method,observationId:observation.id,excerpt:observation.finding,text_pointer:'/observations/'+observationIndex+'/finding',support_rationale:observation.limits}],auditHistory:earlier?[structuredClone(earlier)]:[],cleanup:{reviewRationale:entry.review_rationale,remaining:entry.limits}});
 }
 req(seen.size,'empty cleanup registry is not an additive transition');req([...before].filter(([id,claim])=>equal(claim,after.get(id))).length===before.size-seen.size,'unrelated claim drift');
 model.counts.claim_statuses={};for(const claim of after.values())model.counts.claim_statuses[claim.status]=(model.counts.claim_statuses[claim.status]||0)+1;model.generated_at=registry.generated_at;model.corrections.push(correction(seen.size));
 const artifactHashes=new Map(predecessor.artifactHashes);artifactHashes.set(CLEANUP_PATH,registrySha256);artifactHashes.set(CLEANUP_BASELINE,CLEANUP_BASELINE_SHA256);for(const {artifact,capture} of artifacts.values()){artifactHashes.set(artifact.path,artifact.sha256);artifactHashes.set(capture.source_candidate.path,capture.source_candidate.sha256);for(const support of capture.support_artifacts)artifactHashes.set(support.path,support.sha256);}
 const contextHashes=new Map(predecessor.contextHashes||[]);contextHashes.set(CLEANUP_PATH,registrySha256);contextHashes.set(CLEANUP_BASELINE,CLEANUP_BASELINE_SHA256);for(const {artifact,capture} of artifacts.values()){contextHashes.set(artifact.path,artifact.sha256);contextHashes.set(capture.source_candidate.path,capture.source_candidate.sha256);for(const support of capture.support_artifacts)contextHashes.set(support.path,support.sha256);}
 const matches=new Map([...predecessor.matches].map(([id,match])=>[id,{...match,claim:after.get(id),cleanup:registry.transitions.find(entry=>entry.claim_id===id)||null}]));
 return {...predecessor,model,presentation,matches,artifactHashes,contextHashes,cleanup:{registry,registrySha256,baseline:CLEANUP_BASELINE,baselineSha256:CLEANUP_BASELINE_SHA256,artifacts:[...artifacts.values()].map(({artifact})=>artifact),resultRef:registry.result_ref}};
}

export function loadReviewCleanup({read,model,exists}){
 const present=exists?exists(CLEANUP_PATH):(()=>{try{read(CLEANUP_PATH);return true}catch{return false}})();const marked=(model.corrections||[]).some(item=>item.id===CLEANUP_MARKER);
 if(!present){req(!marked,'cleanup-marked model has no registry');return null}const projected=projectReviewCleanup({read,exists});req(equal(model,projected.model),'whole model differs from cleanup projection');return projected;
}

File: scripts/current_host_review_cleanup.py
#!/usr/bin/env python3
"""Fail-closed editorial cleanup projection above the jurisdiction baseline.

This layer records only claim-specific, repository-local editorial checks.  It
does not execute Host operations or convert a source classification alone into
product evidence.
"""
from __future__ import annotations

import copy
import hashlib
import importlib.util
import json
from collections import Counter
from pathlib import Path
from typing import Any

REGISTRY = 'verification/current-host-review-cleanup.json'
# Set to the frozen registry digest when the independently reviewed proposal is
# sealed.  An empty value is permitted only while no cleanup registry exists.
REGISTRY_SHA256 = 'ae2c8781c9d0d806818649052ad9bace4e8e6e9cf0b08e61f1d797049c20906c'
ATTEMPT = 'verification/evidence/2026-09-11-host-review-cleanup-attempt-01'
BASELINE = ATTEMPT + '/before-current-host-docs-review.json'
BASELINE_SHA256 = 'f18e61882f50a2785f2fe863aba88ecf3d79af8a45bdcbf1596e6e64c6b7059d'
MARKER = 'HOST-REVIEW-CLEANUP-01'
DECISION = 'CURRENT_HOST_REVIEW_CLEANUP'
STATIC = 'REPOSITORY_STATIC_CHECK'
EDITORIAL = {'EDITORIAL_SCOPE_OR_LEAD_IN', 'EDITORIAL_NAVIGATION_PREAMBLE'}
STATIC_CLASSES = {'REVIEWED_ADVICE', 'REVIEWED_NAVIGATION_INSTRUCTION', 'REVIEWED_EXAMPLE_CALCULATION'}
METHODS = {'MARKUP_OR_PREAMBLE_INSPECTION', 'CONTEXTUAL_INSPECTION', 'LOCAL_NAVIGATION_CHECK', 'ARITHMETIC_CHECK', 'MIXED_REPOSITORY_LOCAL_EDITORIAL_CHECKS'}
AFTER = {'classification', 'status', 'required_evidence_types', 'owner_role', 'rationale', 'next_action', 'evidence_refs', 'source_refs'}


def require(ok: Any, message: str) -> None:
    if not ok:
        raise ValueError('cleanup: ' + message)


def digest(value: bytes) -> str:
    return hashlib.sha256(value).hexdigest()


def objhash(value: Any) -> str:
    return digest(json.dumps(value, sort_keys=True, separators=(',', ':'), ensure_ascii=False).encode())


def module(name: str):
    spec = importlib.util.spec_from_file_location(name, Path(__file__).with_name(name + '.py'))
    result = importlib.util.module_from_spec(spec)
    assert spec.loader
    spec.loader.exec_module(result)
    return result


def safe(root: Path, ref: str) -> Path:
    require(isinstance(ref, str) and ref and not ref.startswith('/') and '\\' not in ref and all(p not in {'', '.', '..'} for p in ref.split('/')), 'unsafe path')
    path = root.resolve()
    for part in ref.split('/'):
        path /= part
        require(not path.is_symlink(), 'unsafe symlink')
    require(path.is_file() and path.resolve().is_relative_to(root.resolve()), 'missing path ' + ref)
    return path


def pinned(root: Path, ref: str, wanted: str) -> bytes:
    value = safe(root, ref).read_bytes()
    require(digest(value) == wanted, 'digest drift ' + ref)
    return value


def decode(value: bytes) -> dict[str, Any]:
    result = json.loads(value)
    require(isinstance(result, dict), 'JSON object required')
    return result


def keys(value: Any, expected: set[str]) -> None:
    require(isinstance(value, dict) and set(value) == expected, 'missing/unknown fields')


def predecessor(root: Path, baseline: dict[str, Any]) -> None:
    jurisdiction = module('current_host_jurisdiction')
    jurisdiction.validate_model(baseline, root)


def correction(count: int) -> dict[str, str]:
    return {
        'id': MARKER,
        'scope': f'{count} exact editorial/navigation/calculation cleanup transitions',
        'history': 'The sealed jurisdiction predecessor is hash-pinned. Original claim text, spans, statuses, sources and evidence remain available.',
        'current': f'{count} individually observed repository-local editorial checks; no Host, API, SSH, credentialed, paid or mutating operation was run.',
        'reason': 'A completed navigation, arithmetic or contextual editorial inspection is limited to the documented passage. It does not establish product behavior, a runtime result, policy enforcement or human acceptance.',
    }


def project(root: Path) -> dict[str, Any]:
    root = root.resolve()
    require(REGISTRY_SHA256, 'cleanup registry digest not sealed')
    registry_bytes = pinned(root, REGISTRY, REGISTRY_SHA256)
    registry_sha256 = REGISTRY_SHA256
    registry = decode(registry_bytes)
    keys(registry, {'schema_version', 'record_type', 'generated_at', 'baseline', 'artifacts', 'transitions', 'result_ref'})
    require(registry['schema_version'] == '1.0' and registry['record_type'] == 'HOST_REVIEW_CLEANUP_TRANSITION', 'wrong registry type')
    require(registry['baseline'] == {'path': BASELINE, 'sha256': BASELINE_SHA256}, 'baseline substitution')
    require(isinstance(registry['result_ref'], str) and registry['result_ref'].startswith(ATTEMPT + '/') and safe(root, registry['result_ref']).is_file(), 'invalid result reference')
    baseline = decode(pinned(root, BASELINE, BASELINE_SHA256))
    predecessor(root, baseline)

    artifacts: dict[str, tuple[dict[str, Any], dict[str, Any]]] = {}
    for artifact in registry['artifacts']:
        keys(artifact, {'id', 'path', 'sha256', 'method'})
        aid = artifact['id']
        require(isinstance(aid, str) and aid and aid not in artifacts and artifact['method'] in METHODS, 'invalid artifact')
        require(isinstance(artifact['path'], str) and artifact['path'].startswith(ATTEMPT + '/'), 'artifact outside cleanup attempt')
        capture = decode(pinned(root, artifact['path'], artifact['sha256']))
        keys(capture, {'schema_version', 'record_type', 'generated_at', 'baseline', 'method', 'limits', 'source_candidate', 'support_artifacts', 'observations'})
        require(capture['schema_version'] == '1.0' and capture['record_type'] == 'HOST_REVIEW_CLEANUP_LOCAL_CHECKS', 'wrong cleanup capture')
        require(capture['baseline'] == registry['baseline'] and capture['method'] == artifact['method'], 'capture baseline/method drift')
        require(isinstance(capture['limits'], str) and capture['limits'].strip(), 'capture limits required')
        require(isinstance(capture['observations'], list), 'capture observations required')
        keys(capture['source_candidate'], {'path', 'sha256'})
        candidate = decode(pinned(root, capture['source_candidate']['path'], capture['source_candidate']['sha256']))
        require(candidate.get('record_type') == 'HOST_EDITORIAL_INSPECTION_CANDIDATES' and candidate.get('baseline') == registry['baseline'] and isinstance(candidate.get('records'), list), 'raw independent candidate drift')
        require(isinstance(capture['support_artifacts'], list), 'support artifacts required')
        for support in capture['support_artifacts']:
            keys(support, {'path', 'sha256'})
            pinned(root, support['path'], support['sha256'])
        artifacts[aid] = (artifact, capture, {item.get('claim_id'): item for item in candidate['records']})

    before = {claim['id']: claim for page in baseline['pages'] for claim in page['claims']}
    result = copy.deepcopy(baseline)
    after = {claim['id']: claim for page in result['pages'] for claim in page['claims']}
    seen: set[str] = set()
    allowed_statuses = {'PASS', 'NOT_APPLICABLE'}
    for entry in registry['transitions']:
        keys(entry, {'claim_id', 'before_sha256', 'artifact_id', 'observation_id', 'after', 'review_rationale', 'limits'})
        cid = entry['claim_id']
        require(cid in before and cid not in seen and entry['before_sha256'] == objhash(before[cid]), 'claim identity/predecessor drift')
        seen.add(cid)
        prior, claim = before[cid], after[cid]
        require(prior['status'] == 'UNVALIDATED', 'only known selected UNVALIDATED claims may transition')
        keys(entry['after'], AFTER)
        claim.update(copy.deepcopy(entry['after']))
        require(claim['status'] in allowed_statuses | {'UNVALIDATED'}, 'only PASS, NOT_APPLICABLE or scoped UNVALIDATED cleanup status')
        require(claim['text'] == prior['text'] and claim['headings'] == prior['headings'] and claim['spans'] == prior['spans'], 'source literal/span mutation')
        require(all(ref in claim['evidence_refs'] for ref in prior['evidence_refs']) and all(ref in claim['source_refs'] for ref in prior['source_refs']), 'prior evidence/source dropped')
        require(entry['artifact_id'] in artifacts, 'missing cleanup artifact')
        artifact, capture, candidates = artifacts[entry['artifact_id']]
        observation = next((item for item in capture['observations'] if item.get('id') == entry['observation_id']), None)
        keys(observation, {'id', 'claim_id', 'before_sha256', 'text', 'text_sha256', 'source_file', 'source_sha256', 'spans', 'method', 'result', 'finding', 'limits', 'context', 'expected', 'observed', 'links', 'arithmetic', 'candidate_record'})
        require(prior['spans'] and all(span['source_file'] == prior['spans'][0]['source_file'] for span in prior['spans']), 'cleanup spans must be in one exact source file')
        source_file = prior['spans'][0]['source_file']
        require(observation['claim_id'] == cid and observation['before_sha256'] == entry['before_sha256'] and observation['text'] == prior['text'] and observation['text_sha256'] == digest(prior['text'].encode()), 'observation claim identity drift')
        require(observation['source_file'] == source_file and observation['spans'] == prior['spans'] and observation['source_sha256'] == digest(safe(root, source_file).read_bytes()), 'observation source/span drift')
        require(observation['method'] in METHODS and artifact['method'] == 'MIXED_REPOSITORY_LOCAL_EDITORIAL_CHECKS' and observation['finding'] == entry['review_rationale'] and observation['limits'] == entry['limits'], 'observation rationale/limits drift')
        require(observation['candidate_record'] == candidates.get(cid), 'raw candidate record substitution')
        current_source = safe(root, source_file).read_text()
        require(isinstance(observation['context'], str) and observation['context'] in current_source and isinstance(observation['expected'], str) and observation['expected'].strip() and isinstance(observation['observed'], str) and observation['observed'].strip(), 'raw contextual expected/observed evidence missing')
        require(isinstance(observation['links'], list), 'navigation evidence must be a list')
        require(isinstance(entry['review_rationale'], str) and entry['review_rationale'].strip() and isinstance(entry['limits'], str) and entry['limits'].strip(), 'rationale/limits required')
        if claim['status'] == 'NOT_APPLICABLE':
            require(claim['classification'] in EDITORIAL and claim['required_evidence_types'] == [] and observation['method'] == 'MARKUP_OR_PREAMBLE_INSPECTION' and observation['result'] == 'NOT_APPLICABLE' and observation['arithmetic'] is None, 'invalid editorial non-claim closure')
        else:
            require(claim['classification'] in STATIC_CLASSES and claim['required_evidence_types'] == [STATIC] and observation['method'] in {'CONTEXTUAL_INSPECTION', 'LOCAL_NAVIGATION_CHECK', 'ARITHMETIC_CHECK'} and observation['result'] == ('PASS' if claim['status'] == 'PASS' else 'UNVALIDATED'), 'invalid bounded static closure')
            require(any(ref.get('artifact_ref') == artifact['path'] for ref in claim['evidence_refs']), 'selected static evidence missing')
            if observation['method'] == 'LOCAL_NAVIGATION_CHECK':
                require(observation['links'] and observation['arithmetic'] is None and all(isinstance(item, dict) and set(item) == {'href', 'destination', 'result'} and item['result'] == 'PASS' for item in observation['links']), 'exact local navigation observation required')
            elif observation['method'] == 'ARITHMETIC_CHECK':
                require(not observation['links'] and isinstance(observation['arithmetic'], dict) and set(observation['arithmetic']) == {'inputs', 'formula', 'expected', 'observed'} and observation['arithmetic']['expected'] == observation['arithmetic']['observed'], 'exact arithmetic observation required')
            else:
                require(not observation['links'] and observation['arithmetic'] is None, 'contextual inspection must not pose as link/arithmetic proof')
            if claim['status'] == 'UNVALIDATED':
                require(observation['method'] == 'CONTEXTUAL_INSPECTION' and claim['rationale'] != prior['rationale'] and claim['next_action'] != prior['next_action'], 'unvalidated cleanup needs exact current rationale and next action')
        claim['history'] = {**prior['history'], 'carry_decision': DECISION, 'reason': prior['history'].get('reason', '') + ' Exact repository-local cleanup observation; no product/runtime or human-acceptance result is inferred.'}
        claim['coverage_state'] = 'CHANGED'
    require(seen, 'empty cleanup registry is not an additive transition')
    require(len(before) - sum(before[cid] == after[cid] for cid in before) == len(seen), 'unrelated claim drift')
    result['counts']['claim_statuses'] = dict(sorted(Counter(item['status'] for item in after.values()).items()))
    result['generated_at'] = registry['generated_at']
    result['corrections'].append(correction(len(seen)))
    return result


def validate_model(model: dict[str, Any], root: Path) -> None:
    require(model == project(root), 'whole model differs from cleanup projection')


def load_cleanup(root: Path, model: dict[str, Any]) -> dict[str, Any] | None:
    present = (root / REGISTRY).is_file()
    marked = any(item.get('id') == MARKER for item in model.get('corrections', []))
    if not present:
        require(not marked, 'cleanup-marked model has no registry')
        return None
    validate_model(model, root)
    registry = decode(pinned(root, REGISTRY, REGISTRY_SHA256))
    return {'registry': registry, 'registry_sha256': REGISTRY_SHA256}
